### PR TITLE
Fix dynamic heights for iOS Flyout Item Templates

### LINKED ISSUE #8290: [Bug] [iOS] [Shell] Flyout Cells Don't Fit DataTemplate Content
### Description
The [UIContainerView](https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Platform.iOS/Renderers/UIContainerCell.cs) in the [ShellTableViewSource](https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Platform.iOS/Renderers/ShellTableViewSource.cs) always remains a static size, regardless of the contents height. It appears that this occurs for both MenuItems and ShellItems.

### Steps to Reproduce

1. Open the application
2. View the flyout.
3. Each of the items in the flyout should be a different height due to the HeightRequest of the content in the cell, but each remains a single static height.

### Expected Behavior
Each cell would resize to fit the height of the content inside it (Similar to Android).

### Actual Behavior
The cell is a single static size.

### Basic Information

- Version with issue: 4.2, 4.3
- IDE: Visual Studio for Mac, Visual Studio
- Platform Target Frameworks: 
  - iOS: iOS 12.2 iOS 13.1

### Screenshots
Screen Shot 2019-10-29 at 9 26 01 AM
Screen Shot 2019-10-29 at 9 26 33 AM

### Reproduction Link
[FlyoutCellHeightBug.zip](https://github.com/xamarin/Xamarin.Forms/files/3785678/FlyoutCellHeightBug.zip)

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change ###
- Calculate the height of the templates used on the flyout so that they will size correctly
- Expose the creation of the TableViewController and TableViewSource for the flyout so users can override these and supply custom implementations
- This also fixes a regression on iOS where currently the flyout is unscrollable if content is too large

### Issues Resolved ### 

- fixes #8290 

### Platforms Affected ### 
- iOS

### Behavioral/Visual Changes ###
The height behavior will change on iOS if the templates that have been supplied have a HeightRequest specified or if they have a height > 44. This makes the behavior match Android and it is more expected

### Testing Procedure ###
- compare store shell with and without this change to make sure results makes sense
- verify the menu item with a templated height has specified height
- verify android and ios are consistent

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: Xamarin.Forms.Platform.iOS/Renderers/ShellTableViewController.cs [hunk 1]
```diff
@@ -59,22 +66,16 @@ void UpdateVerticalScrollMode()
...
-					var firstCellIsVisible = TableView.Bounds.Contains(firstCellRect);
-
-					var lastRowIndex = NMath.Max(0, TableView.NumberOfRowsInSection(0) - 1);
-					var pathToLastRow = Foundation.NSIndexPath.FromRowSection(lastRowIndex, 0);
-					var cellRect = TableView.RectForRowAtIndexPath(pathToLastRow);
-					var lastCellIsVisible = TableView.Bounds.Contains(cellRect);
-
-					TableView.ScrollEnabled = !firstCellIsVisible || !lastCellIsVisible;
+					TableView.ScrollEnabled = true;
+					TableView.AlwaysBounceVertical = false;
```
Comment (PureWeen) [inline Xamarin.Forms.Platform.iOS/Renderers/ShellTableViewController.cs]: @paymicro is this ok?   From what I can tell your changes broke scrolling when set to Auto

File: Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs [hunk 2]
```diff
@@ -18,16 +18,23 @@ public class ShellFlyoutContentRenderer : UIViewController, IShellFlyoutContentR
...
 		public ShellFlyoutContentRenderer(IShellContext context)
 		{
+			_shellContext = context;
+
 			var header = ((IShellController)context.Shell).FlyoutHeader;
 			if (header != null)
 				_headerView = new UIContainerView(((IShellController)context.Shell).FlyoutHeader);
-			_tableViewController = new ShellTableViewController(context, _headerView, OnElementSelected);
+
+			_tableViewController = CreateShellTableViewController();
```
Comment (mjfreelancing) [inline Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs]: Is there any concern with calling a virtual method from the constructor?

Comment (PureWeen) [reply Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs]: Yea best practices dictate that I should move this :-) 

I think moving this to a lazy instantiation + wiring this up inside an override of *LoadView* would be the better way to put this together

Comment (mjfreelancing) [reply Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs]: Sounds like a much better plan

File: Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs [hunk 3]
```diff
@@ -18,16 +18,23 @@ public class ShellFlyoutContentRenderer : UIViewController, IShellFlyoutContentR
...
 
 			AddChildViewController(_tableViewController);
 
 			context.Shell.PropertyChanged += HandleShellPropertyChanged;
 
-			_shellContext = context;
+		}
+
+		protected virtual ShellTableViewController CreateShellTableViewController()
```
Comment (PureWeen) [inline Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs]:
```suggestion
		protected ShellTableViewController CreateShellTableViewController()
```

File: Xamarin.Forms.Platform.iOS/Renderers/ShellTableViewController.cs [hunk 4]
```diff
@@ -29,6 +31,11 @@ public ShellTableViewController(IShellContext context, UIContainerView headerVie
 			_context.Shell.PropertyChanged += OnShellPropertyChanged;
 		}
 
+		protected virtual ShellTableViewSource CreateShellTableViewSource()
```
Comment (PureWeen) [inline Xamarin.Forms.Platform.iOS/Renderers/ShellTableViewController.cs]:
```suggestion
		protected ShellTableViewSource CreateShellTableViewSource()
```

**Comment by PureWeen** [comment]
> If a sub-class calls this method what can the caller do with the returned ShellTableViewController / ShellTableViewSource in order to change behavior? Or are you exposing the method for some other reason?

If you're commenting about this commit

https://github.com/xamarin/Xamarin.Forms/pull/8531/commits/11c5bd55d4b645aafca8ff9060d3bad278f8ef28

Virtual was just removed because we've stopped accepting any API changes to 4.4 so I'll make those virtual again in 4.5

> If a sub-class calls this method what can the caller do with the returned ShellTableViewController / ShellTableViewSource in order to change behavior

A subclass wouldn't really call these methods. Once these are made virtual again then a subclass would override them to provide subclassed implementations of *ShellTableViewController / ShellTableViewSource* which are much easier to customize if you can utilize a custom type. You can't customize much with an  iOS view/viewcontroller/class if you don't control the type

**Comment by mjfreelancing** [comment]
Ah ok, I didn't realize the removal of virtual was temporary - makes sense now - that's what raised my eyebrows :-)

**Pull request merged**